content/sd/0005-replicacao-e-tolerancia-a-faltas.md

```diff
@@ -397,10 +397,380 @@ concorrentemente e nenhum consiga a maioria (_deadlock_). Para resolver este pro
+
+**A grande maioria dos sistemas de propagação epidémica aplica as atualizações de
+forma a respeitar a relação [_aconteceu-antes_](/sd/tempo-e-sincronizacao/#eventos-e-relógios-lógicos), ou seja, respeita a ordem
+causal dos acontecimentos.**
+
+E no caso dos **clientes móveis**?
+
+- Um cliente muda a sua _password_ na réplica R1
+- Muda para a réplica R2 antes da alteração ser propagada
+- **Não consegue utilizar a sua nova password**
+
+Existem sistemas capazes de garantir a relação _aconteceu-antes_ para clientes móveis,
+enquanto que outros oferecem **garantias ainda mais fracas** como as
+[garantias de sessão](#garantias-de-sessão) apresentadas acima.
+
+Todas estas garantias podem ser atingidas com o uso de [relógios
+lógicos](/sd/tempo-e-sincronizacao/#eventos-e-relógios-lógicos) (pouco eficiente).
+
+### _Lazy Replication_ - _gossip architecture_
+
+Foi proposta em 1922 e inspirou muitos dos sistemas **fracamente coerentes** dos
+dias de hoje. Tem como objetivo:
+
+- garantir **acesso rápido** aos clientes
+- mesmo quando existem **partições na rede**
+- com o _trade-off_ de sacrificar a **coerência**
+
+O nome _**gossip**_ vem de as réplicas propagarem as alterações periodicamente em
+_background_, como se fossem boatos.
+
+![lazy replication basic operation](./assets/0005-lazy-replication.png#dark=3)
+
+O algoritmo oferece **coerência fraca**, mas assegura duas propriedades:
+
+- _**Monotonic reads**_
+  - mesmo que o cliente aceda a réplicas diferentes
+- O estado das réplicas **respeitam a ordem causal** das alterações
+  - se uma modificação $m_2$ depende de outra $m_1$, a réplica nunca executa $m_2$
+    antes de $m_1$
+
+#### Interação cliente-réplica
+
+- Cada cliente mantém um _timestamp_ vetorial denominado $prev$
+
+  - vetor de inteiros, um por réplica
+  - reflete a última versão acedida pelo cliente
+
+- Em cada contacto com uma réplica, é enviado também o $prev$: (pedido, $prev$)
+- A réplica devolve (resposta, $new$)
+
+  - em que $new$ é o _timestamp_ vetorial que reflete o estado da réplica
+  - se a réplica estiver atrasada em relação ao cliente, espera até se atualizar
+    para devolver uma resposta (caso seja uma leitura)
+
+- Cliente atualiza $prev$ de acordo com $new$, confrontando cada entrada de
+  $prev$ com a correspondente de $new$
+
+  - se $new[i] \gt prev[i]$, $prev[i] = new[i]$
+
+- Cada réplica mantém também localmente um _**update log**_
+  - uma réplica pode já ter recebido uma modificação mas não a poder executar
+    devido a ainda não ter recebido/executado dependências causais, pelo que a
+    mantém em estado pendente no _log_
+  - pode assim propagar modificações individuais para as outras réplicas, mantendo
+    o _update_ no _log_ até receber confirmação de todas elas
+
+![lazy replication basic operation](./assets/0005-lazy-replication-interaction.png#dark=3)
+
+Quando uma réplica recebe um **pedido de leitura**:
+
+- verifica se $pedido.prev \leq value~timestamp$
```

Comment:
```suggestion
- verifica se $\text{pedido.prev} \leq \text{value timestamp}$
```

I think it would make more sense to use the `\text` or `\op` operators here, what do you think?
Same applies for stuff below.

Reply:
We didn't use `\text` because we wanted those keywords to refer to the variables that each replica maintains locally.

Also, why should we use `\op` instead of `\leq`? Is there any particular reason (we are not familiar with KaTeX)?

Reply:
`\op` is a custom shorthand (see `gatsby-config.js`) for `\operatorname`, which would be a replacement for `\text`, not `\leq`.

I think it's just a matter of looking better, since by default I'd say KaTeX/LaTeX treats text as individual letter variables (i.e. `cos` vs `\cos`).

Reply:
We just thought that the default font used by KaTeX suited better this kind of expression, since we want the keywords to refer to pseudo algorithm variables, and with the `\text` wrapper we feel that this idea isn't transmitted to the reader. But if it is preferable to use the `\text` wrapper to mantain overall consistency across the website, we can change it, no problem.

Default font:

![image](https://github.com/leic-pt/resumos-leic/assets/48092637/e1c565bc-6466-4c3c-b19f-7703e207a0e8)
Wrapped with `\text`

![image](https://github.com/leic-pt/resumos-leic/assets/48092637/0b622c0a-cc59-4dcc-8bc3-012bbe1b687e)

Reply:
Fair enough, you can keep it :)

I'd still say `\text` is more semantically correct, but it doesn't "stand out" as much.